source/_posts/2017/2017-10-09-how-to-use-doctrine-repository-as-service-in-symfony.md

```diff
@@ -0,0 +1,438 @@
+    
+    ```php
+    $postRepository = $entityManager->getRepository(Post::class);
+    $postRepository->getPostsByAuthor(); // method not found, or was it found*()?
+    ```
+
+- When we use Doctrine's EntityRepository method, like `get($id)`, we will face same problem:
+ 
+    ```php
+    $post = $this->postRepository->get(1);
+    $post->getName(); # or was it "$post->getTitle()";
+    ```
+    
+To enable typehtins, we have to add them manually.
+
+    ```php
+    /** @var App\Entity\Post $post */
+    $post = $this->postRepository->get(1);
+    $post->getName(); # or was it "$post->getTitle()";
+    ```
+
+This annotation helper should never be in *your* code, except this case:
+ 
+    ```php
+    /** @var SomeService $someService */
+    $someService = $container->get(SomeService::class);
+    ```
+
+
+### 4. Registration `services.yml`
+
+
+None. Everything is handled by Doctrine Repository locator.
+
+
+
+### Advantages Summary
+
+- Everybody does that.
+- It's easy to copy-paste if already present in our code.
+- It's spread in most of documentation, both in Doctrine and Symfony and in many posts about Doctrine.
+- No brain, no gain.
+
+
+### Problems Summary
+
+- We cannot use autowiring.
+- We cannot inject repository to other service just via constructor.
+- We have to **typehint manually** everything (plugins an other workarounds put asside).
+- **We have Doctrine in our Controller** - Controller should only delegate to model, without knowing what Database package is used.  
+- To allow constructor injection, we have to prepare for *config programming*.
+- Thus it's coupled to the framework you use and less reusable.
+- We cannot use multiple repository for single entity. It naturally leads to huge repositories.
+- We cannot use construtor injection in repositoreis, which can easily lead you to creating static helper classes.
+- Also, you directly depend on Doctrine's API, so if `get()` changes to `find()` in one `composer update`, your app is down.
+ 
+
+
+#1## How to Make This Better with Symfony 3.3?
+
+It require few steps, but all leads to single one SOLID change. Have you heard about *composition over inheritance*?
+
+Instead of *inheritance*...
+
+```php
+<?php declare(strict_types=1);
+
+namespace App\Repository;
+
+use App\Entity\Post;
+use Doctrine\ORM\EntityRepository; 
+
+final class PostRepository extends EntityRepository
+{
+}
+```
+
+...we will use *composition*:
+
+
+```php
+<?php declare(strict_types=1);
+
+namespace App\Repository;
+
+use App\Entity\Post;
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\EntityRepository;
+
+final class PostRepository
+{
+    /**
+     * @var Repository  
+     */
+    private $repository;
+    
+    public function __construct(EntityManager $entityManager)
+    {
+        $this->repository = $entityManager->getRepository(Post::class);
```

Comment:
I remember a suggestion (not sure where I read or heard it) that EntityRepository should be avoided even in this cases. With just injecting EM and using DQL instead of findBy() and other repository methods.

I'm not 100% sure about this but it sounds like a good suggestion to me. Just something for you to consider.

Reply:
I think it would be too much for this post. But maybe an idea for your post on Péhápkaři? ;)